tools/file_packager.py

```diff
@@ -475,13 +487,35 @@ def was_seen(name):
     basename = os.path.basename(filename)
     if file_['mode'] == 'embed':
       # Embed
-      data = base64_encode(utils.read_binary(file_['srcpath']))
-      code += '''var fileData%d = '%s';\n''' % (counter, data)
-      code += ('''Module['FS_createDataFile']('%s', '%s', decodeBase64(fileData%d), true, true, false);\n'''
-               % (dirname, basename, counter))
+      binary = utils.read_binary(file_['srcpath'])
+      if not wasmfs_c:
+        # JS output
+        data = base64_encode(binary)
+        code += '''var fileData%d = '%s';\n''' % (counter, data)
+        code += ('''Module['FS_createDataFile']('%s', '%s', decodeBase64(fileData%d), true, true, false);\n'''
+                 % (dirname, basename, counter))
+      else:
+        # C output.
+        # convert the binary data into a C escaped string, \xAB for hex code AB
+        def escape_for_c(char):
+          if char < 16:
+            # add a 0 to keep the output in two bytes
+            return '\\x0' + hex(char)[-1:]
+          return '\\x' + hex(char)[-2:]
+        data = ''.join([escape_for_c(char) for char in binary])
+        # directories... :( make them
+        c_output += f'''static const char fileData{counter}[] = "{data}";\n'''
+        c_output += f'''
+FILE* file{counter} = fopen("{dirname}" "/" "{basename}", "wb");
+fwrite(fileData{counter}, 1, {len(binary)}, file{counter});
```

Comment:
If I'm not mistaken, doing it this way causes the file data to be loaded into memory as part of the data section, and then copied into different memory (assuming this is a memory-backed FS), which seems unfortunate. Here are a couple of crazy ideas about how to avoid this:

1. Somehow put the data into a passive segment, and then drop it after copying. I guess that would probably have to involve assembly or compiler builtins, but might be worth it.
2. Put the data into a custom section instead of the data section. This has the downside that we'd have to go out to JS to get it (and I guess copy it temporarily into wasm memory during initialization). Also I'm not actually clear whether the VM would retain the custom section in its memory or not...

I'm not familiar with how this was done before. Did the old way of embedding avoid this problem somehow? Could we get the data into a JS arraybuffer which could eventually be dropped and GC'ed. I assume that downloading the data from the server rather than having it be embedded isn't an option?

Reply:
Yes, this has overhead - a copy, as you said.

The old method copies the data from JS base64 encoded text during startup, so is even worse...

A passive segment or a custom section is an interesting idea. But I was actually hoping we can use an internal WasmFS API, basically giving it a pointer and telling it that it "owns" the data. Then the data would simply stay where it is. That's a little odd if it's written to (modifying "static" data) but it is as efficient as it gets, and for this purpose it is safe since the data corresponds 1:1 to a file. But maybe there's a problem with that idea?

Reply:
I don't think we need any kind of passive segment or special segment do we?   Can't we just use an internal API and pass a pointer and size?

Reply:
Ah right, avoiding copying at all would be ideal. I guess that would work up until there's any kind of append, then you'd have to allocate more space and copy the whole thing. But for read-only it would be great.

Reply:
Oh, and wrt memory overhead: The data section is part of the wasm file (some of which the engine keeps... somewhere? maybe in memory?), and it's obviously initialized into wasm memory in the data area, and then into wherever it's copied if applicable. Coming from JS, it would be in a JS file (which... again, we don't really know whether the VM keeps that), and once wherever it was copied if applicable (but no extra copy in the wasm memory). So it could actually be less overhead to have it come from JS if we aren't able to re-use the static data as the single copy (e.g. if we think it will be written, or if we are writing into a persistent FS rather than a memory FS).

Reply:
I think we should probably consider writable preloaded files to be out of scope, no?   I can't imagine many folks wanting that... and if they did then it seems reasonable to make a copy of first write.. but aside from that I think a single copy if wasm memory sounds like a good goal.

Reply:
> But I was actually hoping we can use an internal WasmFS API, basically giving it a pointer and telling it that it "owns" the data. Then the data would simply stay where it is.

It would be pretty easy to have a `StaticMemoryFile` that just contains a pointer to some static memory in the data section. It could easily be mutable, too, just like any non-const global byte array in C. Growing it could allocate the additional data in a vector without copying the original data; there's no need for the data to be contiguous.

Regarding the delivery mechanism; all our data segments (both passive and active) are dropped on startup, either manually in the case of passive segments or automatically in the case of active segments. So there shouldn't be any extra copies of the data sitting around. Using custom sections, OTOH, would require an extra copy to be kept somewhere on the JS heap since the Wasm API exposes their contents.